src/librustc_mir/hair/cx/expr.rs

```diff
@@ -111,7 +112,7 @@ fn apply_adjustment<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
                 span,
                 kind: ExprKind::Borrow {
                     region: deref.region,
-                    borrow_kind: to_borrow_kind(deref.mutbl),
+                    borrow_kind: to_borrow_kind(deref.mutbl, true),
```

Comment:
I think I expected us to push these flags even further back, to type check. I'm a bit wary that we may introduce adjustments for reasons that are not "autoref-autoderef in conjunction with method calls".

For example, does this type-check?

```rust
fn foo(x: &mut u32, y: u32) {
    *x += y;
}

fn bar(x: &mut u32) {
    foo(x, *x)
}
```

I think it might, because we get an implicit `&mut *x`  on the call, and that might be treated as two-phase. Similarly `let x: &mut u32 = x` or something like that.

Reply:
(Actually, I suppose we could conceivably thread a bit of state down during HAIR lowering to indicate whether the expression being coerced is in receiver position.)

Reply:
I will figure out how to tighten the restriction. (As you have pointed out there may be different paths to getting there.)

Reply:
(Though of course the example you give *is sound*; its just code that might surprise some if it passes, right? And anyway I agree that for our first deployment we should reject it)

Reply:
Yes, just surprising